File: plateform/frontend/src/api/assets.ts
import { apiClient } from "./client";
import type {
  AssetDetail,
  AssetFilters,
  AssetHealthResponse,
  AssetIncidentListResponse,
  AssetListResponse,
} from "../types/assets";

export function toAssetQuery(filters: AssetFilters): Record<string, string> {
  const query: Record<string, string> = {};
  if (filters.search.trim()) {
    query.search = filters.search.trim();
  }
  if (filters.asset_type) {
    query.asset_type = filters.asset_type;
  }
  if (filters.status) {
    query.status = filters.status;
  }
  if (filters.zone) {
    query.zone = filters.zone;
  }
  if (filters.maintenance) {
    query.maintenance = filters.maintenance;
  }
  if (filters.sort_by) {
    query.sort_by = filters.sort_by;
  }
  if (filters.sort_order) {
    query.sort_order = filters.sort_order;
  }
  return query;
}

export async function fetchAssets(filters: AssetFilters): Promise<AssetListResponse> {
  const { data } = await apiClient.get<AssetListResponse>("/api/assets", {
    params: toAssetQuery(filters),
  });
  return data;
}

export async function fetchAsset(assetId: string): Promise<AssetDetail> {
  const { data } = await apiClient.get<AssetDetail>(
    `/api/assets/${encodeURIComponent(assetId)}`,
  );
  return data;
}

export async function fetchAssetIncidents(
  assetId: string,
): Promise<AssetIncidentListResponse> {
  const { data } = await apiClient.get<AssetIncidentListResponse>(
    `/api/assets/${encodeURIComponent(assetId)}/incidents`,
  );
  return data;
}

export async function fetchAssetHealth(
  assetId: string,
  range?: "1h" | "6h" | "24h" | "7d",
  config: { signal?: AbortSignal } = {},
): Promise<AssetHealthResponse> {
  const { data } = await apiClient.get<AssetHealthResponse>(
    `/api/assets/${encodeURIComponent(assetId)}/health`,
    { params: range ? { range } : undefined, signal: config.signal },
  );
  return data;
}

File: plateform/backend/app/api/routes/assets.py
from fastapi import APIRouter, Depends, Query

from app.api.deps import get_asset_service, get_maintenance_service
from app.schemas.assets import (
    AssetDetail,
    AssetHealthResponse,
    AssetIncidentListResponse,
    AssetListResponse,
    AssetSortField,
    AssetType,
    MaintenanceFilter,
    OperationalStatus,
)
from app.schemas.incidents import SortOrder
from app.schemas.maintenance import AssetMaintenanceResponse
from app.schemas.telemetry import TelemetryRange
from app.services.assets import AssetService
from app.services.maintenance import MaintenanceService

router = APIRouter(prefix="/assets", tags=["assets"])


@router.get("", response_model=AssetListResponse)
def list_assets(
    asset_type: AssetType | None = Query(default=None),
    status: OperationalStatus | None = Query(default=None),
    zone: str | None = Query(default=None),
    search: str | None = Query(default=None),
    maintenance: MaintenanceFilter | None = Query(default=None),
    sort_by: AssetSortField = Query(default=AssetSortField.health_score),
    sort_order: SortOrder = Query(default=SortOrder.asc),
    service: AssetService = Depends(get_asset_service),
) -> AssetListResponse:
    return service.list_assets(
        asset_type=asset_type,
        status=status,
        zone=zone,
        search=search,
        maintenance=maintenance,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.get("/{asset_id}", response_model=AssetDetail)
def get_asset_detail(
    asset_id: str,
    service: AssetService = Depends(get_asset_service),
) -> AssetDetail:
    return service.get_asset(asset_id)


@router.get("/{asset_id}/incidents", response_model=AssetIncidentListResponse)
def get_asset_incidents(
    asset_id: str,
    service: AssetService = Depends(get_asset_service),
) -> AssetIncidentListResponse:
    return service.get_related_incidents(asset_id)


@router.get("/{asset_id}/maintenance", response_model=AssetMaintenanceResponse)
def get_asset_maintenance(
    asset_id: str,
    service: MaintenanceService = Depends(get_maintenance_service),
) -> AssetMaintenanceResponse:
    return service.asset_maintenance(asset_id)


@router.get("/{asset_id}/health", response_model=AssetHealthResponse)
def get_asset_health(
    asset_id: str,
    range: TelemetryRange = Query(default=TelemetryRange.twenty_four_hours),
    service: AssetService = Depends(get_asset_service),
) -> AssetHealthResponse:
    return service.get_health(asset_id, range.value)
